src/sm.c

```diff
@@ -0,0 +1,331 @@
+        /*Codes_SRS_SM_02_009: [ sm_open_begin shall set b_now to 0, succeed and return SM_EXEC_GRANTED. ]*/
+        LONG64 b_now = InterlockedCompareExchange64(&sm->b_now, 0, -1);
+        if (b_now != -1)
+        {
+            /*Codes_SRS_SM_02_012: [ If sm_open_end doesn't follow a call to sm_open_begin then sm_open_end shall return. ]*/
+            LogError("cannot execute, name=%s, b_now=%" PRId64 "", sm->name, b_now);
+            result = SM_EXEC_REFUSED;
+        }
+        else
+        {
+            InterlockedIncrement64(&sm->n);
+            result = SM_EXEC_GRANTED;
+        }
+    }
+    return result;
+}
+
+void sm_open_end(SM_HANDLE sm)
+{
+    if (sm == NULL)
+    {
+        /*Codes_SRS_SM_02_010: [ If sm is NULL then sm_open_end shall return. ]*/
+        LogError("invalid argument SM_HANDLE sm=%p", sm);
+    }
+    else
+    {
+        /*Codes_SRS_SM_02_011: [ sm_open_end shall increment e, b_now to 0 and return. ]*/
+        InterlockedIncrement64(&sm->e);
+        InterlockedExchange64(&sm->b_now, 0);
+    }
+}
+
+/*aims at getting a stable "e" for a stable "n" and returns the difference between them*/
+static LONG64 get_n_minus_e(SM_HANDLE sm)
+{
+    LONG64 e1, e2, n1, n2;
+    do
+    {
+        e1 = InterlockedAdd64(&sm->e, 0);
+        do
+        {
+            n1 = InterlockedAdd64(&sm->n, 0);
+            e2 = InterlockedAdd64(&sm->e, 0);
+            n2 = InterlockedAdd64(&sm->n, 0);
+        } while (n1 != n2);
+    } while (e1 != e2);
+
+    return n1 - e1;
+}
+
+/*sm_barrier_begin calls this. Also sm_close calls this, because _close is nothing else but another barrier*/
+static SM_RESULT sm_barrier_begin_internal(SM_HANDLE sm)
+{
+    SM_RESULT result;
+    LONG64 b_now = InterlockedOr64(&sm->b_now, 1);
+
+    /*Codes_SRS_SM_02_014: [ sm_close_begin shall set lowest bit of b_now to 1. ]*/
+    /*Codes_SRS_SM_02_028: [ If b_now has least significand bit set to 1 then sm_barrier_begin shall fail and SM_EXEC_REFUSED. ]*/
+    /*Codes_SRS_SM_02_020: [ If there was no sm_open_begin/sm_open_end called previously, sm_close_begin shall fail and SM_EXEC_REFUSED. ]*/
+    if ((b_now & 1) == 0)
+    {
+        InterlockedExchange(&sm->e_done, 0);
+
+        LONG64 n = InterlockedIncrement64(&sm->n);
+        (void)n;
+
+        /*Codes_SRS_SM_02_030: [ sm_barrier_begin shall succeed and return SM_EXEC_GRANTED. ]*/
+        /*Codes_SRS_SM_02_017: [ sm_close_begin shall succeed and return SM_EXEC_GRANTED. ]*/
+        result = SM_EXEC_GRANTED;
+
+        /*drain previous executions*/
+        /*Codes_SRS_SM_02_016: [ sm_close_begin shall wait for all previous operations to end. ]*/
+        /*Codes_SRS_SM_02_029: [ sm_barrier_begin shall wait for the completion of all the previous operations. ]*/
+        while (get_n_minus_e(sm) != 1)
+        {
+            if (InterlockedHL_WaitForValue(&sm->e_done, 1, INFINITE) != INTERLOCKED_HL_OK)
+            {
```

Comment:
So this part here is waiting for "pending requests to get to 0". It can be named "the difference between n and e is 1" but essentially that what it does. Honestly the problem we had with the pending request count was that there was no check before incrementing it if someone wants to take the barrier.
This is basically a smarted SRL lcock (because "SRW locks are neither fair nor FIFO." - from MSDN)

So maybe modelling this as a count + a lock bit is actually not that bad - for sure I find it more readable when we need to debug ...
I think it's worth talking about this as we plan to use this in all our code - tomorrow morning maybe?